ssv/spectest/tests/validator/test.go

```diff
@@ -0,0 +1,128 @@
+package validator
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"testing"
+
+	"github.com/attestantio/go-eth2-client/spec/phase0"
+	"github.com/bloxapp/ssv-spec/ssv"
+	"github.com/bloxapp/ssv-spec/types"
+	"github.com/bloxapp/ssv-spec/types/testingutils"
+	"github.com/stretchr/testify/require"
+)
+
+type ValidatorTest struct {
+	Name                   string
+	Duties                 []*types.Duty
+	Messages               []*types.SSVMessage
+	OutputMessages         []*types.SSVMessage
+	BeaconBroadcastedRoots []string
+	ExpectedError          string
+}
+
+func (test *ValidatorTest) TestName() string {
+	return "validator " + test.Name
+}
+
+// Run as an individual test
+func (test *ValidatorTest) Run(t *testing.T) {
+
+	ks := testingutils.Testing4SharesSet()
+
+	v := testingutils.BaseValidator(ks)
+
+	// Init all duty runners
+	v.DutyRunners[types.BNRoleAttester] = testingutils.AttesterRunner(ks)
+	v.DutyRunners[types.BNRoleAggregator] = testingutils.AggregatorRunner(ks)
+	v.DutyRunners[types.BNRoleProposer] = testingutils.ProposerRunner(ks)
+	v.DutyRunners[types.BNRoleSyncCommittee] = testingutils.SyncCommitteeRunner(ks)
+	v.DutyRunners[types.BNRoleSyncCommitteeContribution] = testingutils.SyncCommitteeContributionRunner(ks)
+	v.DutyRunners[types.BNRoleValidatorRegistration] = testingutils.ValidatorRegistrationRunner(ks)
+	v.DutyRunners[types.BNRoleVoluntaryExit] = testingutils.VoluntaryExitRunner(ks)
+
+	v.Network = v.DutyRunners[types.BNRoleAttester].GetNetwork()
+
+	var lastErr error
+
+	// Start each requested duty
+	for _, duty := range test.Duties {
+		lastErr = v.StartDuty(duty)
+	}
+
+	// process each message and store the last error
+	for _, msg := range test.Messages {
+		err := v.ProcessMessage(msg)
+		if err != nil {
+			lastErr = err
+		}
+	}
+
+	// Check error
+	if len(test.ExpectedError) != 0 {
+		require.EqualError(t, lastErr, test.ExpectedError)
+	} else {
+		require.NoError(t, lastErr)
+	}
+
+	// test output message
+	test.compareOutputMsgs(t, v)
+
+	// test beacon broadcasted msgs
+	test.compareBroadcastedBeaconMsgs(t, v)
+}
+
+func (test *ValidatorTest) compareBroadcastedBeaconMsgs(t *testing.T, v *ssv.Validator) {
+
+	broadcastedRoots := make([]phase0.Root, 0)
+
+	for _, role := range []types.BeaconRole{types.BNRoleAttester, types.BNRoleAggregator,
+		types.BNRoleProposer,
+		types.BNRoleSyncCommittee, types.BNRoleSyncCommitteeContribution,
+		types.BNRoleValidatorRegistration, types.BNRoleVoluntaryExit} {
+
+		broadcastedRoots = append(broadcastedRoots, v.DutyRunners[role].GetBeaconNode().(*testingutils.TestingBeaconNode).BroadcastedRoots...)
+	}
+
+	require.Len(t, broadcastedRoots, len(test.BeaconBroadcastedRoots))
+	for _, r1 := range test.BeaconBroadcastedRoots {
+		found := false
+		for _, r2 := range broadcastedRoots {
+			if r1 == hex.EncodeToString(r2[:]) {
+				found = true
+				break
+			}
+		}
+		require.Truef(t, found, "broadcasted beacon root not found")
+	}
```

Comment:
same issue as below I guess

Reply:
Before it I also check the length. But indeed I should add an "already used" map as in the other check. Adding it.